src/network/networkd-dhcp4.c

```diff
@@ -1318,7 +1318,14 @@ int dhcp4_configure(Link *link) {
         if (r < 0)
                 return log_link_warning_errno(link, r, "DHCP4 CLIENT: Failed to set callback: %m");
 
-        r = sd_dhcp_client_set_request_broadcast(link->dhcp_client, link->network->dhcp_broadcast);
+        dhcp_broadcast = link->network->dhcp_broadcast;
+        if (dhcp_broadcast < 0) {
+                if (streq(link->driver, "qeth_l3"))
+                        dhcp_broadcast = true;
+                else
+                        dhcp_broadcast = false;
+        }
```

Comment:
grgrrrr, i really dislike such hw specific hacks.

Here's my proposal for this: let's embed this into hwdb (or possibly just a udev rule). i.e. we already have `20-net-ifname.hwdb` for picking the network interface name for special purpose hw devices (so far only used for some Dell servers, but s390 might actually be a candidate for better named ifaces, too). Let's build on this, and make the default setting of the dhcp broadcast field dependent on hwdb/udev too. This would be a lot more extensible: we can add the s390 case there, but also any other similar kind of hw that needs this.

In general: we want networkd code to be generic and as driver/hw independent as possible. hwdb otoh is our approach for carrying hw-specific tweaks. This is one such hw-specific quirk.

i.e. the code here should use `sd_device_get_property_value(link->sd_device, "ID_NET_DEFAULT_DHCP_BROADCAST", &…)` or so, to retrieve whatever hwdb/udev set there and use that.

Reply:
I like that, as it will make it discoverable. I realized that udev info / networkctl status / systemctl show didn't actually show the setting for RequestBroadcast. Thus testing this code was tricky. I did packet dumps and added extra logging in this file.

Re: names - s390x network interface are already stable, predictable and short based on the stable CCW bus ids. That was done by me and other people for all the s390x networking devices.

Reply:
@xnox well, what i wanted to suggest is that sometimes special hw like servers have some very specific ethernet ports or other kind of networking ports, and it might make sense to name them for their actual purpose, if there's a fixed number of them, instead of numeric/opaque bus IDs. 

i.e. if I would build a server hw and i'd have 5 ethernet ports on the back and 7 on the front, it might make sense to use hwdb to name the fron ports front0, front1, front2, … and the back ones back0, back1, back2.. or so. It's a lot nicer than "ens0" os so...

Or alternatively, if they are color coded use "red0", "green0", blue"0"… you get the idea.

i have never seen an s390 device. maybe they have some fixed ethernet devices where such pretty names might be a good idea, don't know

Reply:
@xnox while it is true that channel based network interfaces on s390x have stable names, we are in fact currently working on improving the situation for native PCI based NICs on s390x. See PR #19017 and also my current LKML thread on adding a `/sys/bus/pci/devices/<dev>/index` attribute on s390x here: https://lore.kernel.org/linux-pci/20210412135905.1434249-1-schnelle@linux.ibm.com/

So I think @mihajlov and I would be quite interested if there are any efforts to work on hwdb based special interface naming rules or if any one of you has comments on the `../index` proposal. We're definitely open for suggestions.

I think as an architecture which always runs virtualized we definitely have some special requirements though interestingly I find that the virtualized nature sometimes makes Linux on s390x requirements closer to KVM or VMWare guests than classic server platforms. For example none of our network devices are really fixed to a particular OS instance and can always hotplug.

Reply:
So i'm not sure how hwdb can ever work in this case. This is a qeth device, if one binds it with qeth_l2 driver it will be in L2 mode and accept dhcp just fine. When it is in L3 it needs broadcast at the moment.

There is no modalias for this. And the same qeth can change from one driver to another.

Reply:
hmm, are you saying there are two alternative drivers for the same device? how do you switch between the two?

also, yuck!

Reply:
@poettering to begin with no ccw devices are activated at all. Hence one specifies a very long kernel commandline, or activates each device by hand. And then appropriate drive is bound..... There is no automatic activation of any devices by default. Because all of them could be shared across multiple LPARs and multiple z/VMs with them wanted exclusive access to things. Hence it is normal for many LPARs to see many offline devices, and only online some of them. And bind them to appropriate driver. Hence there is a lot of s390-tools tooling to do that with like lszdev / chzdev / lsqeth / etc.